proposed/disable-multi-level-lookup-by-default.md

```diff
@@ -0,0 +1,204 @@
+
+I personally don't like either option. 2 is too restrictive and might make
+the message effectively useless. 1 feels counterproductive as it basically goes
+back to current state where setting `DOTNET_MULTILEVEL_LOOKUP=0` is very common.
+3 feels even worse as it introduces yet another env. variable to set.
+
+### Hosting APIs accept install location as parameter
+
+Currently most hosting APIs take a parameter which specifies the "dotnet root",
+so that they can behave as if one used `dotnet` from that location.
+But the APIs themselves are implemented in `hostfxr` which can be of any version.
+This is already a niche problem (if a really old `hostfxr` is used to run things
+on a really new private install the behavior is a bit problematic).
+With this change such discrepancy would become more apparent.
+So the question is to which version should the APIs react to:
+
+1. Use the version of the `hostfxr` itself - this would be similar to the above
+   described behavior of reacting to the version of the `dotnet.exe` itself,
+   but it might be a bit unexpected.
+1. Use the version of the specified "dotnet root" - this would be more in line
+   with what the "dotnet root" parameter tries to convey. This might be technically
+   a bit challenging in some cases.
+
+I think it should be OK to go with option 1 - rely on the version of `hostfxr` -
+but this will need more detailed investigation.
+
+## Alternative solutions
+
+### Off by default everywhere
+
+In the above, there's a discrepancy between application and SDK behavior.
+SDK will always have multi-level lookup disabled by default,
+even if the projects asks for a lower version (via `global.json`).
+On the other hand apps, which ask for a lower version (via `.runtimeconfig.json`)
+will get a backward compatible behavior - multi-level lookup on by default.
+
+It might be a good idea to make the behavior consistent and always treat
+multi-level lookup as off by default. The impact would be that trying to run
+applications targeting .NET 5 or lower will NOT run if executed on
+a private install of .NET 6.
+
+* Pros - consistent behavior across all use cases
+* Cons - more breaking behavior
+
+SDK is generally backward compatible on its own, that's why we always pick
+the latest version available if not specified otherwise. `global.json` is
+the override mechanism which lets users specify more precise version requirements.
+That said having `global.json` and also using a private install location is
+typically a very explicit action, and so it's reasonable to assume users
+will keep these consistent.
+
+Doing the same for applications is a tougher proposition because they're not
+governed by one `global.json` (typically repos or solutions have only one `global.json`),
+instead they're governed by their own `.runtimeconfig.json` which may
+(and usually are) targeting multiple versions even from one project/repo.
+
+### Off by default everywhere and auto-roll-forward
+
```

Comment:
I'm afraid that this would cause the same kind of problems that we run into now with MLL.  Users won't know that they've been upgraded, and if something doesn't work quite right, there won't be an immediately obvious way to show them that they've been upgraded.

Reply:
I think a variant of this is the winning ticket:

- .NET 7+ doesn't honor MLL.
- There are good controls to enable manual roll-forward of apps (including global controls), with nothing automatic.

Reply:
Sounds great.  Thanks @richlander.  Do we need to make any changes to this doc, or do you think this is ready for merge?

Reply:
I agree with @richlander  on this. For applications we're almost there (missing exact version override, but roll forward works everywhere). For SDK we should add commandline/env options.

I think that if we pay close attention to the experience when things fail - so our error messages/docs, there should be much less need to make things work magically as well as up-front messages.